dashboard/app/models/contact_rollups_raw.rb

```diff
@@ -23,25 +23,44 @@ def self.truncate_table
   end
 
   def self.extract_email_preferences
-    query = extract_from_source_query('email_preferences', ['opt_in'], 'email')
+    query = get_extraction_query('email_preferences', false, ['opt_in'], 'email')
     ActiveRecord::Base.connection.execute(query)
   end
 
-  # @param source [String] Source from which we want to extract data
+  def self.extract_parent_emails
+    source_sql = <<~SQL
+      SELECT parent_email, MAX(updated_at) AS updated_at
+      FROM users
+      GROUP BY parent_email
+    SQL
+
+    query = get_extraction_query(source_sql, true, [], 'parent_email', "#{CDO.dashboard_db_name}.users.parent_email")
+    ActiveRecord::Base.connection.execute(query)
+  end
+
+  # @param source [String] Source from which we want to extract data (can be a dashboard table name, or subquery)
+  # @param source_is_subquery [Boolean] True if source is a subquery, rather than a table name
   # @param data_columns [Array] Columns we want reshaped into a single JSON object
   # @param email_column [String] Column in source table we want to insert ino the email column
+  # @param source_name [String] Name for source (should be non-nil if using a subquery or non-dashboard table)
   # @return [String] A SQL statement to extract and reshape data from the source table.
-  def self.extract_from_source_query(source, data_columns, email_column)
+  def self.get_extraction_query(source, source_is_subquery, data_columns, email_column, source_name=nil)
+    if source_name.nil? && source_is_subquery
+      raise "Source name required if source is a subquery"
+    end
+
+    wrapped_source, sources_column = format_source(source, source_is_subquery, source_name)
```

Comment:
You can keep `get_extraction_query` a bit simpler by removing one of its responsibilities, which is prefixing database name to source name.

Let the caller explicitly add database name to the source table, such as:
`query = get_extraction_query("#{CDO.dashboard_db_name}.email_preferences", false, ['opt_in'], 'email')`

Then this part could be simplified as:
```suggestion
    source_name ||= source
    wrapped_source = source_is_subquery ? "(#{source}) AS subquery"  : source
```

And then you can use `source_name` directly as the value for `sources` column.

Reply:
SO much better, thank you!